### PR TITLE
Cleanup and type annotations

### PR DESCRIPTION
**Pull request by thomasZen**
- Remove a few unnecessary functions and arguments
- Added type annotations
- Fixed bug (see comments)

I will test if it's still running with Lara's model.

### CODE REVIEW

File: openretina/hoefling_2024/models.py [hunk 1]
```diff
@@ -31,16 +31,11 @@ def __repr__(self):
...
 
 class Core3d(Core):
-    def initialize(self, cuda=False):
+    def initialize(self):
         self.apply(self.init_conv)
-        self.put_to_cuda(cuda=cuda)
-
-    def put_to_cuda(self, cuda):
-        if cuda:
-            self = self.cuda()
```
Comment (thomasZen) [inline openretina/hoefling_2024/models.py:33]: This happens automatically if we call model.to(device)

File: openretina/optimization/objective.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 class ResponseReducer(ABC):
     def __init__(self, axis: int = 0):
-        self._axis = 0
+        self._axis = axis
```
Comment (thomasZen) [inline openretina/optimization/objective.py:7]: This was a bug

File: openretina/training.py [hunk 3]
```diff
@@ -149,8 +149,8 @@ def full_objective(model, data_key, inputs, targets, detach_core):
             if (batch_no + 1) % optim_step_count == 0:
                 optimizer.step()
                 optimizer.zero_grad()
-        if np.isnan(loss.item()):
-            raise ValueError(f"Loss is NaN on batch {batch_no} from {data_key}, stopping training.")
+            if np.isnan(loss.item()):
+                raise ValueError(f"Loss is NaN on batch {batch_no} from {data_key}, stopping training.")
```
Comment (thomasZen) [inline openretina/training.py:153]: This was also a bug, as we only raised the ValueError if the last loss was a nan, but we want to raise it if any loss was nan.

File: openretina/hoefling_2024/models.py (hunk 1 above)
Comment (fededagos) [reply openretina/hoefling_2024/models.py:33]: Did you test this? I agree that having a put_to_cuda method does not make sense, but it seems that Core is not inheriting from nn.Module right now so .to(device) might not work

File: openretina/training.py (hunk 3 above)
Comment (fededagos) [reply openretina/training.py:153]: Good catch!

File: openretina/hoefling_2024/models.py (hunk 1 above)
Comment (fededagos) [reply openretina/hoefling_2024/models.py:33]: But given that I have never used that method myself and only .to(device) I think it should work already

Comment (thomasZen) [reply openretina/hoefling_2024/models.py:33]: You're right that model.to(device) doesn't work as the class doesn't inherit from nn.Module.
For the same reason .to(device) doesn't work:
```
>>> from openretina.hoefling_2024.models import Core3d
>>> c3d = Core3d()
>>> c3d.cuda()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Core3d' object has no attribute 'cuda'
```
I think actually the put_to_cuda function was never actively used.

I will go ahead and let Core inherit from nn.Module, but still get rid of the push_to_cuda method.

**Review by fededagos — APPROVED** [review]
Looks all good to me!

**Pull request merged**